packages/block-library/src/navigation/edit.js

```diff
@@ -126,56 +182,181 @@ function Navigation( {
 	}
 
 	function handleCreateFromExistingPages() {
-		updateNavItemBlocks( defaultPagesNavigationItems );
+		updateNavItemBlocks( navLinkBlocksFromPages );
 		selectBlock( clientId );
 	}
 
-	const hasPages = hasResolvedPages && pages && pages.length;
+	function handleCreateFromExistingMenu() {
+		updateNavItemBlocks( navLinkBlocksFromMenuItems );
+		selectBlock( clientId );
+	}
 
-	const blockInlineStyles = {
-		fontSize: fontSize.size ? fontSize.size + 'px' : undefined,
-	};
+	function handleCreate() {
+		const { key } = selectedDropDownOption;
+
+		// Explicity request to create empty.
+		if ( key === CREATE_EMPTY_OPTION_VALUE ) {
+			return handleCreateEmpty();
+		}
+
+		// Create from Pages.
+		if ( hasPages && key === CREATE_FROM_PAGES_OPTION_VALUE ) {
+			return handleCreateFromExistingPages();
+		}
+
+		// Create from WP Menu (if exists and not empty).
+		if (
+			hasMenus &&
+			selectedDropDownOption &&
+			navLinkBlocksFromMenuItems?.length
+		) {
+			return handleCreateFromExistingMenu();
+		}
+
+		// Default to empty menu
+		return handleCreateEmpty();
+	}
+
+	function buildPlaceHolderInstructionText() {
+		if ( hasMenus && hasPages ) {
+			return __(
+				'Create a Navigation from all existing pages, or chose a Menu.'
+			);
+		}
+
+		if ( ! hasMenus && ! hasPages ) {
+			return __( 'Create a an empty Navigation.' );
+		}
+
+		if ( hasMenus && ! hasPages ) {
+			return __( 'Create a Navigation from a Menu or create empty.' );
+		}
+
+		if ( ! hasMenus && hasPages ) {
+			return __(
+				'Create a Navigation from all existing pages, or create empty.'
+			);
+		}
+	}
+
+	const dropDownOptions = [
+		{
+			id: CREATE_PLACEHOLDER_VALUE,
+			name: __( 'Select where to start from…' ),
+		},
+		...( hasMenus ? menus : [] ),
+		{
+			id: CREATE_EMPTY_OPTION_VALUE,
+			name: __( 'Create empty menu' ),
+		},
+		...( hasPages
+			? [
+					{
+						id: CREATE_FROM_PAGES_OPTION_VALUE,
+						name: __( 'New from all top-level pages' ),
+					},
+			  ]
+			: [] ),
+	];
+
+	function shouldDisableCreateButton() {
+		// If there is no key at all then disable.
+		if ( ! selectedDropDownOption?.key ) {
+			return true;
+		}
+
+		// Always disable if the default "placeholder" option is selected.
+		if ( selectedDropDownOption.key === CREATE_PLACEHOLDER_VALUE ) {
+			return true;
+		}
+
+		// Always enable if Create Empty is selected.
+		if ( selectedDropDownOption.key === CREATE_EMPTY_OPTION_VALUE ) {
+			return false;
+		}
+
+		// Enable if Pages option selected and we have Pages available.
+		if (
+			selectedDropDownOption.key === CREATE_FROM_PAGES_OPTION_VALUE &&
+			hasResolvedPages
+		) {
+			return false;
+		}
+
+		// Only "menu" options use an integer based key.
+		const selectedOptionIsMenu = Number.isInteger(
+			selectedDropDownOption.key
+		);
+
+		const menuItemsResolved =
```

Comment:
Should the menu items should only be fetched when the `Create` button is clicked? It doesn't feel right to block the button for a reason the user is unaware of.

Reply:
That needs some thought. We'd need to wait to create the block until after the items are fetched. What happens if there are errors during fetch? Lots of consider from this.

Reply:
@talldan I'd really appreciate any help/insight on how we could achieve the "fetch on demand". I know we get request items via the `getMenuItems` select, but then we have to wait on the resolvers to know when it's finished. I'm a little lost right now on how to structure this code.

Reply:
It won't be straightforward. It might be better to handle this in a separate PR, as this one is already quite big.

But for entities it would usually be achieved using the `isResolving`/`hasFinishedResolving` state to show some kind of loading indication.

I think it could work as follows:
- At the moment, placeholder content can only be shown when Menus and Pages have finished resolving, as that's needed to determine the content. A spinner should probably be used in the placeholder while that content is still loaded to avoid the content changing while being displayed (e.g. `isRequestingMenus && isRequestingPages && ( <Spinner /> )`).
- Once the user selects an option that represents a menu in the dropdown we could start optimistically loading menu items for that menu (call `getMenuItems` in `CustomSelectControl`'s `onChange` when the option is a menu). If menu items are still resolving when the user clicks the create button then we can use the `isBusy` prop on the button (similar to the publish button) to indicate something is happening (`isBusy={ isCreatingFromMenu && isRequestingMenuItems }`. I imagine `isCreatingFromMenu` could be some state that's set to `true` when the button is clicked and set to `false` again when the navigation is created:
https://github.com/WordPress/gutenberg/blob/master/packages/components/src/button/index.js#L30
- The final piece would be handling the result, which as a hook might look like this:
```jsx
useEffect( () => {
  if ( isCreatingFromMenu && menuItems?.length ) {
    const blocks = converMenuItemsToBlocks( menuItems );
    onCreate( blocks );
    setIsCreatingFromMenu( false );
  }
}, [ menuItems, isCreatingFromMenu ] );
```

The logic is getting pretty complicated and it's hard to separate the creation code from the block's editing code, so at this point it might be better to move the Navigation block's placeholder into a separate component.

I'd imagine it could offer some callback props like this:
```
<NavigationPlaceholder
    onCreate( ( blocks ) => { // ... } )
/>
```

Reply:
> It might be better to handle this in a separate PR, as this one is already quite big.

Shall we try to tidy up what we have here and merge what we have which should be working even if it's not the ideal user experience?

Then I will:

1. Raise a follow up which will refactor the `<Placeholder />` outside the block. Nothing else. Get it merged.
2. Create a PR to update the creation flow to mirror what you've indicated above in terms of `isBusy` and resolving...etc. Get it merged.
3. Anything else.

What do you think?